# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: extensions/permissions/PermissionDelegateIPCUtils.h
# Commit: 44e715808901
# Full Hash: 44e71580890195d163912b96f41ef09e32570ec8
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2020-06-13 03:21:52
# Regressor Bug: 1587743
# File Overlap Count: 1
# Description:
#   Bug 1587743 - Part 1: Pre-compute the delegated permissions for the top-level content and store it in the WindowContext. r=baku,nika
#   
#   In order to delegate the permission to the top-level window, in this
#   patch, we pre-compute the permissions of the top-level context and set
#   them to the top-level WindowContext. So, the cross-origin iframe can
# ==============================================================================

diff -r a890be633ec3 -r 44e715808901 extensions/permissions/PermissionDelegateIPCUtils.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/extensions/permissions/PermissionDelegateIPCUtils.h	Fri Jun 12 16:31:49 2020 +0000
@@ -0,0 +1,41 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef mozilla_permissiondelegateipcutils_h
+#define mozilla_permissiondelegateipcutils_h
+
+#include "ipc/IPCMessageUtils.h"
+
+#include "mozilla/PermissionDelegateHandler.h"
+
+namespace IPC {
+
+template <>
+struct ParamTraits<
+    mozilla::PermissionDelegateHandler::DelegatedPermissionList> {
+  typedef mozilla::PermissionDelegateHandler::DelegatedPermissionList paramType;
+
+  static void Write(Message* aMsg, const paramType& aParam) {
+    for (auto& permission : aParam.mPermissions) {
+      WriteParam(aMsg, permission);
+    }
+  }
+
+  static bool Read(const Message* aMsg, PickleIterator* aIter,
+                   paramType* aResult) {
+    for (auto& permission : aResult->mPermissions) {
+      if (!ReadParam(aMsg, aIter, &permission)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+};
+
+}  // namespace IPC
+
+#endif  // mozilla_permissiondelegateipcutils_h